### PR TITLE
boards: adafruit_feather_stm32f405: Add SPI flash

### PR DESCRIPTION
**Pull request by rosterloh**
Added SPI flash definition on SPI1.

Signed-off-by: Richard Osterloh <richard.osterloh@gmail.com>

### CODE REVIEW

**Comment by rosterloh** [comment]
Depends on #22603

**Review by erwango — CHANGES_REQUESTED** [review]
Not a spi-flash expert but this looks good in general.
Two requests though:
- Can you update board doc accordingly?
- Can you add board to samples/drivers/spi_flash so it can get compiled during sanitychecks?

File: boards/arm/adafruit_feather_stm32f405/pinmux.c [hunk 1]
```diff
@@ -22,6 +22,16 @@ static const struct pin_config pinconf[] = {
...
+#ifdef CONFIG_SPI_1
+#ifdef CONFIG_SPI_STM32_USE_HW_SS
+	{STM32_PIN_PA15, STM32F4_PINMUX_FUNC_PA15_SPI1_NSS |
+			 STM32_OSPEEDR_VERY_HIGH_SPEED},
+#endif /* CONFIG_SPI_STM32_USE_HW_SS */
+	{STM32_PIN_PB3, STM32F4_PINMUX_FUNC_PB3_SPI1_SCK |
+			STM32_OSPEEDR_VERY_HIGH_SPEED},
+	{STM32_PIN_PB4, STM32F4_PINMUX_FUNC_PA6_SPI1_MISO},
+	{STM32_PIN_PB5, STM32F4_PINMUX_FUNC_PA7_SPI1_MOSI},
+#endif	/* CONFIG_SPI_1 */
```
Comment (FRASTM) [inline boards/arm/adafruit_feather_stm32f405/pinmux.c]: #endif /* CONFIG_SPI_1 */

Comment (rosterloh) [reply boards/arm/adafruit_feather_stm32f405/pinmux.c]: Done

**Comment by rosterloh** [comment]
@erwango This should be ready to go now. Tested with output:
```
JEDEC SPI-NOR SPI flash testing
==========================

Test 1: Flash erase
Flash erase succeeded!

Test 2: Flash write
Attempting to write 4 bytes
Data read matches data written. Good!!
```

File: samples/drivers/spi_flash/boards/adafruit_feather_stm32f405.conf [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+#
+
+CONFIG_SPI_NOR=y
+
+CONFIG_USB=y
+CONFIG_USB_DEVICE_STACK=y
+CONFIG_USB_DEVICE_PRODUCT="Adafruit STM32F405 Feather"
+CONFIG_USB_DEVICE_VID=0x239A
+CONFIG_USB_DEVICE_PID=0x805A
+CONFIG_USB_UART_CONSOLE=y
```
Comment (erwango) [inline samples/drivers/spi_flash/boards/adafruit_feather_stm32f405.conf:14]: Can you elaborate why USB would be needed for this sample ?

Comment (rosterloh) [reply samples/drivers/spi_flash/boards/adafruit_feather_stm32f405.conf:14]: It is not strictly needed but with this particular device the console over USB is much easier for debugging.

Comment (erwango) [reply samples/drivers/spi_flash/boards/adafruit_feather_stm32f405.conf:14]: Ok, then this would apply to all samples and then maybe this config could be applied as default board config (or optional config) and removed from this sample ?

Comment (rosterloh) [reply samples/drivers/spi_flash/boards/adafruit_feather_stm32f405.conf:14]: That is completely up to you. I think it's best to leave it up to the user as they may want to use USB for something else but I personally always use USB CDC console on this board.

Comment (erwango) [reply samples/drivers/spi_flash/boards/adafruit_feather_stm32f405.conf:14]: My point is I don't think that we should modify all samples or tests specifically for this board (that for the 'c' part of the change). Then if configuration is really this is useful, it would be better to centralize it at board level, so it is always available w/o need to add this config to all samples.

Comment (rosterloh) [reply samples/drivers/spi_flash/boards/adafruit_feather_stm32f405.conf:14]: I totally understand. That would be a separate PR anyway so just tell me if you'd like me to take it out for this PR and I'll update it.

File: samples/drivers/spi_flash/src/main.c [hunk 3]
```diff
@@ -24,11 +26,19 @@
...
+#define FLASH_TEST_REGION_OFFSET 0xf000
+#else
 #define FLASH_TEST_REGION_OFFSET 0xff000
+#endif
 #define FLASH_SECTOR_SIZE        4096
 
 void main(void)
 {
+#if defined(CONFIG_BOARD_ADAFRUIT_FEATHER_STM32F405)
+	usb_enable(NULL);
```
Comment (galak) [inline samples/drivers/spi_flash/src/main.c:39]: Why is this needed, please add a comment at least.

File: samples/drivers/spi_flash/boards/adafruit_feather_stm32f405.conf (hunk 2 above)
Comment (erwango) [reply samples/drivers/spi_flash/boards/adafruit_feather_stm32f405.conf:14]: Ok, let's focus this PR on SPI flash and have a new PR for this USB config, this would be good indeed.

Comment (rosterloh) [reply samples/drivers/spi_flash/boards/adafruit_feather_stm32f405.conf]: Done

File: samples/drivers/spi_flash/src/main.c (hunk 3 above)
Comment (rosterloh) [reply samples/drivers/spi_flash/src/main.c]: Removed as per discussion above

File: boards/arm/adafruit_feather_stm32f405/pinmux.c [hunk 4]
```diff
@@ -22,6 +22,16 @@ static const struct pin_config pinconf[] = {
 	{STM32_PIN_PB6, STM32F4_PINMUX_FUNC_PB6_I2C1_SCL},
 	{STM32_PIN_PB7, STM32F4_PINMUX_FUNC_PB7_I2C1_SDA},
 #endif /* CONFIG_I2C_1 */
+#ifdef CONFIG_SPI_1
+#ifdef CONFIG_SPI_STM32_USE_HW_SS
+	{STM32_PIN_PA15, STM32F4_PINMUX_FUNC_PA15_SPI1_NSS |
+			 STM32_OSPEEDR_VERY_HIGH_SPEED},
+#endif /* CONFIG_SPI_STM32_USE_HW_SS */
+	{STM32_PIN_PB3, STM32F4_PINMUX_FUNC_PB3_SPI1_SCK |
```
Comment (FRASTM) [inline boards/arm/adafruit_feather_stm32f405/pinmux.c:30]: text above in the `index.rst`  is mentioning "SPI1 uses SCK (PA5)", but PB3 is set as SCK here. 
is this mismatching ?

Comment (rosterloh) [reply boards/arm/adafruit_feather_stm32f405/pinmux.c:30]: Good catch, updated.

**Review by MaureenHelm — APPROVED** [review]

**Review by erwango — APPROVED** [review]
